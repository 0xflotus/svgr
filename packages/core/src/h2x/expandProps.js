import { JSXAttribute } from 'h2x-plugin-jsx'

const expandProps = (position = 'end') => () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (
          path.node.name === 'svg' &&
          !path.node.attributes.some(attr => attr && attr.name === 'props')
        ) {
          const props = new JSXAttribute()
          props.name = 'props'
          props.spread = true
          if (position === 'start') {
            path.node.attributes = [props, ...path.node.attributes]
            path.replace(path.node)
          }
          if (position === 'end') {
            path.node.attributes = [...path.node.attributes, props]
            path.replace(path.node)
          }
        }
      },
    },
  },
})

export default expandProps
