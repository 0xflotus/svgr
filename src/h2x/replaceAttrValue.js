const replaceAttrValue = (color, newColor) => () => ({
  visitor: {
    JSXAttribute: {
      enter(path) {
        if (path.node.value === color) {
          path.node.value = newColor
        }
      },
    },
  },
})

export default replaceAttrValue
