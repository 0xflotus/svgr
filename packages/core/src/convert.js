import svgo from './plugins/svgo'
import h2x from './plugins/h2x'
import prettier from './plugins/prettier'
import transform from './plugins/transform'
import { expandState } from './util'
import { loadConfig } from './config'

async function convert(code, baseConfig = {}, baseState = {}) {
  const state = expandState(baseState)
  const config = await loadConfig(baseConfig, baseState)
  let result = code
  // Remove null-byte character (copy/paste from Illustrator)
  result = String(result).replace('\0', '')
  result = await svgo(result, config, state)
  result = await h2x(result, config, state)
  result = await transform(result, config, state)
  result = await prettier(result, config, state)
  return result
}

export default convert
