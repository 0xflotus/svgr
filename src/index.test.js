import jsx from 'h2x-plugin-jsx'
import stripAttribute from './h2x/stripAttribute'
import expandProps from './h2x/expandProps'
import replaceAttrValue from './h2x/replaceAttrValue'
import convert from './'

describe('convert', () => {
  it('should convert', async () => {
    const result = await convert(
      `
      <?xml version="1.0" encoding="UTF-8"?>
      <svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
          <title>Dismiss</title>
          <desc>Created with Sketch.</desc>
          <defs></defs>
          <g id="Blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
              <g id="Dismiss" stroke="#063855" stroke-width="2">
                  <path d="M51,37 L37,51" id="Shape"></path>
                  <path d="M51,51 L37,37" id="Shape"></path>
              </g>
          </g>
      </svg>
    `,
      null,
      { filePath: 'MyComponent.js' },
    )

    expect(result).toMatchSnapshot()
  })

  it('should be possible to specify options', async () => {
    const result = await convert(
      `
      <?xml version="1.0" encoding="UTF-8"?>
      <svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
          <title>Dismiss</title>
          <desc>Created with Sketch.</desc>
          <defs></defs>
          <g id="Blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
              <g id="Dismiss" stroke="#063855" stroke-width="2">
                  <path d="M51,37 L37,51" id="Shape"></path>
                  <path d="M51,51 L37,37" id="Shape"></path>
              </g>
          </g>
      </svg>
    `,
      {
        h2x: {
          plugins: [
            jsx,
            stripAttribute('xmlns'),
            expandProps,
            replaceAttrValue('#063855', 'currentColor'),
          ],
        },
      },
      { filePath: 'MyComponent.js' },
    )

    expect(result).toMatchSnapshot()
  })
})
