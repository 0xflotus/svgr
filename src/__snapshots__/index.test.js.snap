// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should be possible to specify options 1`] = `
"import React from \\"react\\";

const MyComponent = props => (
  <svg width={88} height={88} viewBox=\\"0 0 88 88\\" {...props}>
    <title>Dismiss</title>
    <g
      stroke=\\"currentColor\\"
      strokeWidth={2}
      fill=\\"none\\"
      fillRule=\\"evenodd\\"
      strokeLinecap=\\"square\\"
    >
      <path d=\\"M51 37L37 51M51 51L37 37\\" />
    </g>
  </svg>
);

export default MyComponent;
"
`;

exports[`convert should convert 1`] = `
"import React from \\"react\\";

const MyComponent = props => (
  <svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 88 88\\" {...props}>
    <title>Dismiss</title>
    <g
      stroke=\\"#063855\\"
      strokeWidth={2}
      fill=\\"none\\"
      fillRule=\\"evenodd\\"
      strokeLinecap=\\"square\\"
    >
      <path d=\\"M51 37L37 51M51 51L37 37\\" />
    </g>
  </svg>
);

export default MyComponent;
"
`;
